### PR TITLE
Add do not disturb switch

### PR DESCRIPTION
**Pull request by DurgNomis-drol**
This adds a switch where you can toggle Do Not Disturb.

Alarms and timers will still ring but everything else is turned off. Where i live the feature does not seem to do much, but I believe it is very useful in other countries. 

I have tested it locally and it works! Can some one else test it also?

Because we have an update-interval set at 10 secs, the switch can jump back when used, it will switch to the correct positions when we retrieve new information. 😄 Maybe we should change this?

### CODE REVIEW

File: custom_components/google_home/models.py
Comment (KapJI) [inline custom_components/google_home/models.py:89]:
```suggestion
        """Set Do Not Disturb status."""
```

File: custom_components/google_home/models.py
Comment (KapJI) [inline custom_components/google_home/models.py:93]:
```suggestion
        """Return Do Not Disturb status."""
```

File: custom_components/google_home/models.py
Comment (KapJI) [inline custom_components/google_home/models.py:40]: Make it only available through getter.
```suggestion
        self._do_not_disturb = False
```

File: custom_components/google_home/models.py
Comment (KapJI) [inline custom_components/google_home/models.py:88]: Why `str`?
```suggestion
    def set_do_not_disturb_status(self, status: bool) -> None:
```

File: README.md
Comment (KapJI) [inline README.md:109]: Update table of contents.

File: README.md
Comment (KapJI) [inline README.md]:
```suggestion
| Platform | Sample switch                  | Description                                         |
```

File: README.md
Comment (KapJI) [inline README.md:115]: Maybe remove this column completely?
```suggestion
| `switch` | `switch.living_do_not_disturb` | Toggles Do Not Disturb mode on a Google Home device |
```

Comment (KapJI) [inline README.md:115]:
```suggestion
| `sensor` | `switch.living_do_not_disturb` | Toggle "Do Not Disturb" mode on a Google Home device |
```

File: custom_components/google_home/api.py
Comment (KapJI) [inline custom_components/google_home/api.py:206]:
```suggestion
        """Collect data from different endpoints."""
```

**Review by KapJI — COMMENTED** [review]
I'll take a look at the rest later.

File: custom_components/google_home/models.py
Comment (DurgNomis-drol) [reply custom_components/google_home/models.py:88]: Because `mypy` complains about casting a `str` to a bool when we get it from the endpoint in `api.py`. But you are right it should preferably be a bool. 😄

Comment (KapJI) [reply custom_components/google_home/models.py:88]: I think it should be caster in `api.py`. Also be aware that:
```python
>>> bool("false")
True
>>> bool("False")
True
```

Comment (DurgNomis-drol) [reply custom_components/google_home/models.py:88]: I moved it to `api.py` 😄 Sometimes the errors are hard to understand, but it works now.

**Comment by DurgNomis-drol** [comment]
@KapJI  Yeah i had to move somethings around and split others up. Maybe some of the things can be done a little cleaner. Just take you time.

**Comment by KapJI** [comment]
@leikoilja I noticed several times that CI is not run on PRs. Can it be caused by running out of usage limits? Or is it just github issue? Can you check?
https://docs.github.com/en/github/setting-up-and-managing-billing-and-payments-on-github/viewing-your-github-actions-usage

File: custom_components/google_home/api.py
Comment (KapJI) [inline custom_components/google_home/api.py:348]:
```suggestion
                is_enabled = bool(response["notifications_enabled"])
```

File: custom_components/google_home/api.py
Comment (KapJI) [inline custom_components/google_home/api.py:356]:
```suggestion
                device.set_do_not_disturb_status(is_enabled)
```

File: custom_components/google_home/models.py
Comment (KapJI) [inline custom_components/google_home/models.py:88]:
```suggestion
    def set_do_not_disturb(self, status: bool) -> None:
```

File: custom_components/google_home/models.py
Comment (KapJI) [inline custom_components/google_home/models.py:92]:
```suggestion
    def get_do_not_disturb(self) -> bool:
```

File: custom_components/google_home/api.py
Comment (KapJI) [inline custom_components/google_home/api.py:208]:
```suggestion
```

File: custom_components/google_home/api.py
Comment (KapJI) [reply custom_components/google_home/api.py:356]: Does it blink with this?

File: custom_components/google_home/switch.py
Comment (KapJI) [inline custom_components/google_home/switch.py:79]: Create a function to reduce code duplication.
```suggestion
        self._set_status(True)
```

**Review by KapJI — CHANGES_REQUESTED** [review]

File: custom_components/google_home/api.py
Comment (DurgNomis-drol) [reply custom_components/google_home/api.py:331]: It has to bee inverted. When it `notifications_enabled` is `false`, Do Not Disturb is on.

Comment (KapJI) [reply custom_components/google_home/api.py:331]: Well, `not enable` then 🙂

File: custom_components/google_home/switch.py
Comment (DurgNomis-drol) [reply custom_components/google_home/switch.py:79]: Shall i do it for all the sensors as well?

Comment (KapJI) [reply custom_components/google_home/switch.py:79]: I can't find similar duplication in sensors, so probably not.

**Comment by leikoilja** [comment]
@KapJI, hey i been noticing weird github actions behavior today as well on a work github account. As being open source we don't have limits or price, so it seems github issue 
image

**Comment by KapJI** [comment]
Thanks for checking 🙂
I think push will trigger it then.

**Comment by DurgNomis-drol** [comment]
@KapJI I will look into your requested changes tomorrow or Friday. 😃

**Comment by KapJI** [comment]
@leikoilja let's publish the new version in the meantime? It should close issue with BadAuthentication errors at least.

**Comment by KapJI** [comment]
Let's finish this and publish the new version.

**Comment by DurgNomis-drol** [comment]
@KapJI Push incoming

File: custom_components/google_home/switch.py
Comment (KapJI) [inline custom_components/google_home/switch.py:62]:
```suggestion
        """Return true if Do Not Distrurb Mode is on."""
```

File: custom_components/google_home/switch.py
Comment (KapJI) [inline custom_components/google_home/switch.py:70]: I think it should be the opposite. The switch is on when do not disturb mode is on.
```suggestion
        return device.get_do_not_disturb()
```

File: custom_components/google_home/api.py
Comment (KapJI) [reply custom_components/google_home/api.py:221]: You didn't update this.

File: custom_components/google_home/api.py
Comment (KapJI) [inline custom_components/google_home/api.py:349]: I think you shouldn't call `json.dumps` at all.
```python
>>> json.dumps(None)
'null'
```

Typing may be difficult in this case. In `types.py` you can define such type:
```python
JsonDict = Dict[str, Union[bool, int, str, List[str]]
```
It's not the exact type JSON dict may have, but it's enough for all these cases.

And then use `Optional[JsonDict]` for `data` arg of `post()`.

File: custom_components/google_home/switch.py
Comment (KapJI) [inline custom_components/google_home/switch.py:30]:
```suggestion
    switches: List[SwitchEntity] = []
```

**Review by KapJI — CHANGES_REQUESTED** [review]

File: custom_components/google_home/switch.py
Comment (KapJI) [inline custom_components/google_home/switch.py:62]:
```suggestion
        """Return true if Do Not Disturb Mode is on."""
```

File: custom_components/google_home/api.py
Comment (DurgNomis-drol) [reply custom_components/google_home/api.py:349]: Thanks, I tried early on not using `json.dumps` and had problems with `types`. Will test this. 😄

File: custom_components/google_home/switch.py
Comment (DurgNomis-drol) [reply custom_components/google_home/switch.py:70]: That is how it is now. The setting on device endpoint is inverted. Or do you mean that you want the switch to be off when do not disturb is on? 🤔

Comment (KapJI) [reply custom_components/google_home/switch.py:70]: No, the switch should be on, when do not disturb is on.

`device.get_do_not_disturb()` should return `True` when it's on.

This is how you set it, you set `True` to turn on and `False` to turn off.

File: custom_components/google_home/api.py
Comment (KapJI) [inline custom_components/google_home/api.py:361]: Maybe you have some confusion here? Shouldn't it be?
```python
device.set_do_not_disturb(not is_enabled)
```

Comment (KapJI) [reply custom_components/google_home/api.py:361]: This may explain why you have blinking state.

Comment (DurgNomis-drol) [reply custom_components/google_home/api.py:361]: Yes, that was why i was confused about what you where talking about above. Saw this and have changed it now. 😃

Comment (KapJI) [reply custom_components/google_home/api.py:361]: Rename `is_enabled` to `notifications_enabled` to avoid confusion.

File: custom_components/google_home/api.py
Comment (DurgNomis-drol) [reply custom_components/google_home/api.py:349]: I get this error with mypy:

```text
custom_components/google_home/api.py:277: error: Argument "data" to "post" of "GlocaltokensApiClient" has incompatible type "Dict[str, List[str]]"; expected "Optional[Dict[str, Union[bool, int, str, List[str]]]]"
custom_components/google_home/api.py:277: note: "Dict" is invariant -- see http://mypy.readthedocs.io/en/latest/common_issues.html#variance
custom_components/google_home/api.py:277: note: Consider using "Mapping" instead, which is covariant in the value type
custom_components/google_home/api.py:319: error: Argument "data" to "post" of "GlocaltokensApiClient" has incompatible type "Dict[str, str]"; expected "Optional[Dict[str, Union[bool, int, str, List[str]]]]"
custom_components/google_home/api.py:319: note: "Dict" is invariant -- see http://mypy.readthedocs.io/en/latest/common_issues.html#variance
custom_components/google_home/api.py:319: note: Consider using "Mapping" instead, which is covariant in the value type
custom_components/google_home/api.py:350: error: Argument "data" to "post" of "GlocaltokensApiClient" has incompatible type "Optional[Dict[str, bool]]"; expected "Optional[Dict[str, Union[bool, int, str, List[str]]]]"
```

File: README.md
Comment (leikoilja) [inline README.md:116]:
```suggestion
| `switch` | `switch.living_room_do_not_disturb` | Toggle Do Not Disturb mode on a Google Home device |
```

File: custom_components/google_home/api.py
Comment (leikoilja) [inline custom_components/google_home/api.py]: Can these calls be batched into `gather` to be performed in parallel, @KapJI?
That can be a good starting point where we could add more and more parallel calls as we go (hopefully Google Home will open up some more endpoints, like reminders(been requested feature from 2016 :( ))

File: custom_components/google_home/api.py
Comment (KapJI) [reply custom_components/google_home/api.py:349]: Do what it suggests:
```python
JsonDict = Mapping[str, Union[bool, int, str, List[str]]
```

File: custom_components/google_home/api.py
Comment (leikoilja) [reply custom_components/google_home/api.py]: Oh, just saw that `collect_data_from_endpoints` itself is under `gather`, so maybe my thought would not bring much value :)

File: custom_components/google_home/api.py
Comment (leikoilja) [inline custom_components/google_home/api.py:346]: Not a biggie, but it is a good habit to avoid negations if not necessary. It would make it more readable if the logic reversed: 
```suggestion
        if enable:
             data = {"notifications_enabled": not enable}
            _LOGGER.debug(
                "Setting Do Not Disturb setting to %s on Google Home device %s",
                not enable,
                device.name,
            )
        else:
            _LOGGER.debug(
                "Getting Do Not Disturb setting from Google Home device %s",
                device.name,
            )
```

File: custom_components/google_home/api.py
Comment (leikoilja) [inline custom_components/google_home/api.py:352]: Perhaps we could move `"notifications_enabled"` into as we do in other places constants? :)

**Review by KapJI — APPROVED** [review]
Awesome, thanks! 🚀

**Comment by KapJI** [comment]
@leikoilja sorry, I merged it before you posted your comments, let's fix them in another PR 🙂

**Comment by leikoilja** [comment]
No worries :) 
haha, yeh, i was just reading though when saw the label go "Merged" :) 
Just a few suggestions, not really anything critical.
Awesome job otherwise 🚀

File: custom_components/google_home/api.py
Comment (KapJI) [reply custom_components/google_home/api.py]: We can, but I'm afraid that querying the same device in parallel may be troublesome. But we can try.

**Comment by leikoilja** [comment]
Apart from that, you guys did an awesome job with the last two merges. 
Shall we drop a minor release with these fixes? They seem needed 💥

**Comment by KapJI** [comment]
@leikoilja we also have broken options. Should be a quick fix, do you want to take a look? I think it'd be nice to fix it before the next release.

```
File "/home/ruslan/.homeassistant/custom_components/google_home/config_flow.py", line 118, in async_step_user
    self.options.update(user_input)
AttributeError: 'mappingproxy' object has no attribute 'update'
```

**Comment by leikoilja** [comment]
Ou, where has that came from, 2021.4.0? 😅 
I ll take a look later tonight

**Comment by DurgNomis-drol** [comment]
@KapJI You are bit to quick with merging... I got another problem that i am trying to fix with the do no disturb switch.

**Comment by KapJI** [comment]
I think they were always broken but now they're more explicit about that 🙂

**Comment by KapJI** [comment]
@DurgNomis-drol let's continue in the new PR, smaller ones are easier to review :)

**Comment by DurgNomis-drol** [comment]
Yes lets do that, but next time i would prefer if it was me who merged my own PR. 😃

**Comment by KapJI** [comment]
Makes sense, sorry 🙂

**Comment by leikoilja** [comment]
@DurgNomis-drol, but also to be fair, it is usually someone else who merges the PR, almost never the author.
Quite often, the developer who implemented the changes might not even have merge access. 
The author proposes the solution & requests reviews from teammates when it's all done and dusted :)
If there are some uncertainties, like more tests needed or a second opinion on the implementation it's totally fine to mark the PR with WIP or add some additional comments. But once the review is requested and not marked with WIP or any other special cases, tbh i'd also assume that it's up for a final look and merge  😅

**Comment by DurgNomis-drol** [comment]
@leikoilja I know it is that usually, i just didn't not expected him to merge it right away. 😃I have no problem with any of you merging btw my PR's. Sometimes things goes a little to fast here (including myself).

**Pull request merged**
